internal/node/node.go

```diff
@@ -32,88 +31,96 @@ type Node struct {
 	finderAPI *httpfinderserver.Server
 	adminAPI  *adminserver.Server
 	p2pAPI    *p2pfinderserver.Server
+
+	cancelP2pFinder context.CancelFunc
 }
 
 // New creates a new Node process from CLI
-func New(ctx context.Context, cctx *cli.Context) (*Node, error) {
+func New(cacheSize int, dir, storeType, finderAddr, adminAddr string, p2pEnabled bool) (*Node, error) {
 	n := new(Node)
 	var resultCache cache.Interface
 	var valueStore store.Interface
 
-	cacheSize := int(cctx.Int64("cachesize"))
 	if cacheSize != 0 {
 		resultCache = radixcache.New(cacheSize)
 		log.Infow("result cache enabled", "size", cacheSize)
 	} else {
 		log.Info("result cache disabled")
 	}
 
-	storageDir, err := checkStorageDir(cctx.String("dir"))
+	storageDir, err := checkStorageDir(dir)
 	if err != nil {
 		return nil, err
 	}
 
-	storageType := cctx.String("storage")
-	switch storageType {
+	switch storeType {
 	case "sth":
 		valueStore, err = storethehash.New(storageDir)
 	case "prgreb":
 		valueStore, err = pogreb.New(storageDir)
 	default:
-		err = fmt.Errorf("unrecognized storage type: %s", storageType)
+		err = fmt.Errorf("unrecognized store type: %s", storeType)
 	}
 	if err != nil {
 		return nil, err
 	}
-	log.Infow("Value storage initialized", "type", storageType, "dir", storageDir)
+	log.Infow("Value storage initialized", "type", storeType, "dir", storageDir)
 
 	n.indexer = core.NewEngine(resultCache, valueStore)
 	log.Infow("Indexer engine initialized")
-	n.finderAPI, err = httpfinderserver.New(cctx.String("finder_ep"), n.indexer)
+	n.finderAPI, err = httpfinderserver.New(finderAddr, n.indexer)
 	if err != nil {
 		return nil, err
 	}
 	log.Infow("Admin API initialized")
-	n.adminAPI, err = adminserver.New(cctx.String("admin_ep"), n.indexer)
+	n.adminAPI, err = adminserver.New(adminAddr, n.indexer)
 	if err != nil {
 		return nil, err
 	}
 	log.Infow("Client finder API initialized")
-	p2pEnabled := cctx.Bool("enablep2p")
+
 	if p2pEnabled {
+		ctx, cancel := context.WithCancel(context.Background())
+
 		// NOTE: We are creating a new flat libp2p host here because no other
 		// process in the indexer node needs a libp2p host. In the future, when
 		// the indexer node starts using other libp2p protocols to interact with
 		// miners and other indexers, we may need to initialize it before and
 		// use it here so we have a single libp2p host giving service to the whole indexer.
-		h, err := libp2p.New(ctx)
+		p2pHost, err := libp2p.New(ctx)
```

Comment:
Just a reminder. We should create the libp2p host using the keys generated in the config. Let's maybe include it already in this PR so we don't forget in the future.

Also, this libp2p host will be shared by the ingestion and discovery modules of the indexer, so it won't be exclusively initialized with `p2pEnabled` (this is not the case for now).

Reply:
Yes, that willl be in the next PR that combines the config PR and this one.